README.md
# sh lib ssh
## desc

some function to curd ssh for ymc shell lib

## deps

```sh
list=$(ls sh_modules) ; echo "$list" | sed "s/^ */- [x] /g"
list=$(ls sh_modules) ; echo "$list" | sed "s/^ */- [ ] /g"
```

### dev

- [ ] sh-lib-path-resolve.sh
- [x] sh-lib-project-dir-map.sh
- [x] sh-lib-load-lib.sh
- [ ] sh-lib-lang.sh
- [x] sh-lib-gpg.sh

### tes

- [ ] sh-lib-path-resolve.sh
- [x] sh-lib-project-dir-map.sh
- [x] sh-lib-load-lib.sh
- [ ] sh-lib-lang.sh
- [ ] sh-lib-test.sh
- [ ] sh-lib-time-cost.sh

### pro

- [ ] sh-lib-gpg.sh
- [ ] sh-lib-lang.sh
- [ ] sh-lib-load-lib.sh
- [ ] sh-lib-path-resolve.sh
- [ ] sh-lib-project-dir-map.sh
- [ ] sh-lib-project-path-get.sh
- [ ] sh-lib-test.sh
- [ ] sh-lib-this-file.sh
- [ ] sh-lib-time-cost.sh
- [ ] sh-lib-txt-part.sh
- [ ] sh-lib-var-list.sh

### sam

- [ ] sh-lib-path-resolve.sh
- [x] sh-lib-project-dir-map.sh
- [x] sh-lib-load-lib.sh
- [ ] sh-lib-lang.sh


## apis

- [x] ssh_var_config
- [x] ssh_var_caculate
- [x] ssh_secret_key_list
- [x] ssh_secret_key_list_some
- [x] ssh_secret_key_gen
- [x] ssh_secret_key_del
- [x] ssh_secret_key_get
- [x] ssh_key_check
- [x] ssh_pub_key_get
- [x] ssh_pub_key_copy
- [x] ssh_pub_key_upload
- [x] ssh_pub_key_copy_id_to_ms
- [x] ssh_pub_key_copy_id_to_ms_by_expect0
- [x] ssh_pub_key_copy_id_to_ms_by_expect
- [x] ssh_pri_key_get
- [x] ssh_pri_key_push
- [x] ssh_pri_key_add_to_known_host
- [x] ssh_pto_vm
- [x] ssh_pto_vm_expect
- [x] ssh_to_vm
- [x] print_ssh_to_vm
- [x] ssh_all_key_scp_to_ms
- [x] scp_to_vm
- [x] print_scp_to_vm
- [x] ssh_config_file_add

```sh
cat src/index.sh |  grep "function " | sed "s/function */- [x] /g"  | sed "s/(.*) *{//g"
```

## feats

- [x] basic curd ssh
- [x] git push with ssh

## usage

### how to use for poduction?

```sh
# get the code

# run the index file
# ./src/index.sh

# or import to your sh file
# source /path/to/the/index file

# simple usage

```

### how to use for developer?

```sh
# get the code

# run test
# ./test/index.sh
#2 get some fail test
# ./test/index.sh | grep "it is false"
```

## author

yemiancheng <ymc.github@gmail.com>

## license

MIT


CHANGELOG.md
# change log

## desc

starting

## deps

### test

- [ ] sh-lib-path-resolve
- [ ] sh-lib-test
- [ ] sh-lib-time-cost
- [ ] sh-lib-project-dir-map

### prod

- [ ] xxx

## apis

- [x] ssh_var_config
- [x] ssh_var_caculate
- [x] ssh_secret_key_list
- [x] ssh_secret_key_list_some
- [x] ssh_secret_key_gen
- [x] ssh_secret_key_del
- [x] ssh_secret_key_get
- [x] ssh_key_check
- [x] ssh_pub_key_get
- [x] ssh_pub_key_copy
- [x] ssh_pub_key_upload
- [x] ssh_pub_key_copy_id_to_ms
- [x] ssh_pub_key_copy_id_to_ms_by_expect0
- [x] ssh_pub_key_copy_id_to_ms_by_expect
- [x] ssh_pri_key_get
- [x] ssh_pri_key_push
- [x] ssh_pri_key_add_to_known_host
- [x] ssh_pto_vm
- [x] ssh_pto_vm_expect
- [x] ssh_to_vm
- [x] print_ssh_to_vm
- [x] ssh_all_key_scp_to_ms
- [x] scp_to_vm
- [x] print_scp_to_vm
- [x] ssh_config_file_add

## feats

- [x] basic curd ssh
- [x] git push with ssh

test/unit/index.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:test unit

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto

echo "todos:test.unit"

# file usage
# test/unit/index.sh
# time test/unit/index.sh

test/unit/entry02.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

mod_require "sh-lib-gpg.sh" ;
mod_require "${SRC_PATH}/index.sh" ;

echo "todos:test.unit"

# file usage
# test/unit/entry02.sh

sample/unit/index.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:test unit

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto

echo "todos:test.unit"

# file usage
# sample/unit/index.sh

sample/tom/app.del.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:ssh key add

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto


#project_dir_map_out
#project_dir_map_out3
#f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_out3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# ini var
#ssh_var_ini

# set cnf
SSH_SERVER_IP="192.168.56.3"
SSH_USER_NAME=root
SSH_KEY_FILE_NAME=$(date "+%y-%m-%d") #$(date "+%y-%m-%d") #$(date "+%Y-%m-%d")
SSH_PRI_KEY_FILE_PATH=~/.ssh/

# cnf from arg
[ "$1" ] && SSH_SERVER_IP="$1" ; [ -z "$SSH_SERVER_IP" ] && SSH_SERVER_IP="192.168.56.3"
[ "$2" ] && SSH_USER_NAME="$2" ; [ -z "$SSH_USER_NAME" ] && SSH_USER_NAME=root
[ "$3" ] && SSH_KEY_FILE_NAME="$3" ; [ -z "$SSH_KEY_FILE_NAME" ] && SSH_KEY_FILE_NAME=$(date "+%y-%m-%d")
[ "$4" ] && SSH_PRI_KEY_FILE_PATH="$4" ; [ -z "$SSH_PRI_KEY_FILE_PATH" ] && SSH_PRI_KEY_FILE_PATH=~/.ssh/


#ssh_var_config "192.168.56.3" "root" "20-06-16" ~/.ssh/
#ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "2020-10-02" "$SSH_PRI_KEY_FILE_PATH"
ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"
ssh_secret_key_del

# file usage
# sample/tom/app.del.sh
# sample/tom/app.del.sh "192.168.56.3" "root" "20-06-16" ~/.ssh/
# sample/tom/app.del.sh "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"

sample/tom/app.get.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:ssh key add

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto


#project_dir_map_out
#project_dir_map_out3
#f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_out3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# ini var
#ssh_var_ini

# set cnf
SSH_SERVER_IP="192.168.56.3"
SSH_USER_NAME=root
SSH_KEY_FILE_NAME=$(date "+%y-%m-%d") #$(date "+%y-%m-%d") #$(date "+%Y-%m-%d")
SSH_PRI_KEY_FILE_PATH=~/.ssh/

# cnf from arg
[ "$1" ] && SSH_SERVER_IP="$1" ; [ -z "$SSH_SERVER_IP" ] && SSH_SERVER_IP="192.168.56.3"
[ "$2" ] && SSH_USER_NAME="$2" ; [ -z "$SSH_USER_NAME" ] && SSH_USER_NAME=root
[ "$3" ] && SSH_KEY_FILE_NAME="$3" ; [ -z "$SSH_KEY_FILE_NAME" ] && SSH_KEY_FILE_NAME=$(date "+%y-%m-%d")
[ "$4" ] && SSH_PRI_KEY_FILE_PATH="$4" ; [ -z "$SSH_PRI_KEY_FILE_PATH" ] && SSH_PRI_KEY_FILE_PATH=~/.ssh/


#ssh_var_config "192.168.56.3" "root" "20-06-16" ~/.ssh/
#ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "2020-10-02" "$SSH_PRI_KEY_FILE_PATH"
ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"

#ssh_secret_key_list
ssh_secret_key_list_some
#ssh_secret_key_get
ssh_pub_key_get
#ssh_pri_key_get

# file usage
# sample/tom/app.get.sh
# sample/tom/app.get.sh "192.168.56.3" "root" "20-06-16" ~/.ssh/
# sample/tom/app.get.sh "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"


sample/tom/app.use.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:ssh key add

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto


#project_dir_map_out
#project_dir_map_out3
#f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_out3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# ini var
#ssh_var_ini

# set cnf
SSH_SERVER_IP="192.168.56.18"
SSH_USER_NAME=root
SSH_KEY_FILE_NAME=$(date "+%y-%m-%d") #$(date "+%y-%m-%d") #$(date "+%Y-%m-%d")
SSH_PRI_KEY_FILE_PATH=~/.ssh/

# cnf from arg
[ "$1" ] && SSH_SERVER_IP="$1" ; [ -z "$SSH_SERVER_IP" ] && SSH_SERVER_IP="192.168.56.13"
[ "$2" ] && SSH_USER_NAME="$2" ; [ -z "$SSH_USER_NAME" ] && SSH_USER_NAME=root
[ "$3" ] && SSH_KEY_FILE_NAME="$3" ; [ -z "$SSH_KEY_FILE_NAME" ] && SSH_KEY_FILE_NAME=$(date "+%y-%m-%d")
[ "$4" ] && SSH_PRI_KEY_FILE_PATH="$4" ; [ -z "$SSH_PRI_KEY_FILE_PATH" ] && SSH_PRI_KEY_FILE_PATH=~/.ssh/
[ "$5" ] && SSH_SERVER_PORT="$5" ; [ -z "$SSH_SERVER_PORT" ] && SSH_SERVER_PORT=22

#ssh_var_config "192.168.56.18" "root" "20-06-25" ~/.ssh/
#ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "2020-10-02" "$SSH_PRI_KEY_FILE_PATH" $SSH_SERVER_PORT
ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH" $SSH_SERVER_PORT

# prepare for using
#ssh_pub_key_upload
#ssh_pub_key_copy_id_to_ms
#ssh_pub_key_copy_id_to_ms_by_expect0
#ssh_pub_key_copy_id_to_ms_by_expect
#ssh_pri_key_push
#ssh_pri_key_add_to_known_host

# case: tell ssh to login ms with some ip,port,user,key . eg: login ms without -i ssh root@192.168.56.18 -P 22
#

# case: copy key files to some ms
#ssh_all_key_scp_to_ms

# case: copy some files to some ms
#scp_to_vm

# case: login some ms
#ssh_pto_vm
#ssh_pto_vm_expect
ssh_to_vm

#print_ssh_to_vm
#print_scp_to_vm


# file usage
# sample/tom/app.use.sh
# sample/tom/app.use.sh "192.168.56.18" "root" "20-06-25" ~/.ssh/ 22
# sample/tom/app.use.sh "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH" $SSH_SERVER_PORT


sample/tom/app.add.sh
#!/bin/sh

THIS_FILE_PATH=$(cd $(dirname $0);pwd)
THIS_FILE_NAME=$(basename $0)
[ -z "$RUN_SCRIPT_PATH" ] && RUN_SCRIPT_PATH=$(pwd)

#echo "$THIS_FILE_PATH"
#echo "$RUN_SCRIPT_PATH"


# desc:get project dir path
# desc:include lib project-dir-map and(to) gen project dir map
# desc:include lib lib-load and(to) load lib
# desc:ssh key add

# step-x:get project dir path relative to THIS_FILE_PATH
PATH_TO_PROJECT_RELATIVE="../../";PROJECT_PATH=$(cd "$THIS_FILE_PATH";cd "$PATH_TO_PROJECT_RELATIVE" ;pwd)

# step-x:project-dir-map load
source "${PROJECT_PATH}/sh_modules/sh-lib-project-dir-map.sh"
# step-x:gen project dir map with default tpl
#[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; project_dir_map_gen3 "../"
# step-x:gen project dir map with custom tpl from file  or default tpl
[ -z "$THIS_FILE_PATH" ] && THIS_FILE_PATH=$(this_file_path_get) ; f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_gen3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# step-x:sh-lib-load-lib load
source "${PROJECT_PATH}/sh_modules/sh-lib-load-lib.sh"
# step-x:sh-lib-load-lib use

# step-x : lib require and load
#mod_require "${SH_MODULE_LOCAL}/sh-lib-gpg.sh" ;
SH_MODULE_LOCAL="${PROJECT_PATH}/sh_modules"
# step-x : get p dir
SH_MODULE_GLOBAL=$(cd "$PROJECT_PATH";cd "../" ;pwd)
# step-x : lib require only
mod_require_only "sh-lib-gpg.sh" ;
mod_require_only "${SRC_PATH}/index.sh" ;
# step-x : lib load auto
mod_load_auto


#project_dir_map_out
#project_dir_map_out3
#f="${PROJECT_PATH}/P_DIR_MAP.md" ; [ -e $f ] && P_DIR_MAP_LIST=$(cat "${PROJECT_PATH}/P_DIR_MAP.md") ; project_dir_map_out3 "$PATH_TO_PROJECT_RELATIVE" "$P_DIR_MAP_LIST"

# ini var
#ssh_var_ini

# set cnf
SSH_SERVER_IP="192.168.56.3"
SSH_USER_NAME=root
SSH_KEY_FILE_NAME=$(date "+%y-%m-%d") #$(date "+%y-%m-%d") #$(date "+%Y-%m-%d")
SSH_PRI_KEY_FILE_PATH=~/.ssh/

# cnf from arg
[ "$1" ] && SSH_SERVER_IP="$1" ; [ -z "$SSH_SERVER_IP" ] && SSH_SERVER_IP="192.168.56.3"
[ "$2" ] && SSH_USER_NAME="$2" ; [ -z "$SSH_USER_NAME" ] && SSH_USER_NAME=root
[ "$3" ] && SSH_KEY_FILE_NAME="$3" ; [ -z "$SSH_KEY_FILE_NAME" ] && SSH_KEY_FILE_NAME=$(date "+%y-%m-%d")
[ "$4" ] && SSH_PRI_KEY_FILE_PATH="$4" ; [ -z "$SSH_PRI_KEY_FILE_PATH" ] && SSH_PRI_KEY_FILE_PATH=~/.ssh/


#ssh_var_config "192.168.56.3" "root" "20-06-16" ~/.ssh/
#ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "2020-06-16 "$SSH_PRI_KEY_FILE_PATH"
ssh_var_config "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"
ssh_secret_key_gen



function var_list_key_out(){
  echo "$var_list"
}
# fun usage
# var_list_key_out

function var_list_val_out(){
  local key=,line=,list=
  list=$(echo "$var_list" | sed "/^ *$/d")
  echo "$list" | while read line
  do
      key=$(echo "$line" | cut -d "=" -f 1)
      eval "echo $key=\$$key"
  done
}
# fun usage
# var_list_val_out

todo_var_to_update=$(cat<<EOF
$todo_var_to_update
PRIVITE_KEY_FILE_NAME=SSH_PRI_KEY_FILE_NAME
PRIVITE_KEY_FILE_PATH=SSH_PRI_KEY_FILE_PATH
SSH_SERVER_IP=
SSH_SERVER_USER=SSH_USER_NAME
PUBLIC_KEY_FILE_NAME=SSH_PUB_KEY_FILE_NAME
var_to_add=SSH_SERVER_PORT
var_to_add=SSH_USER_NOTE
var_to_add=SSH_USER_MAIL
var_to_add=SSH_KEY_PASSPHRASE
var_to_add=SSH_USER_PASSPHRASE
var_to_add=SSH_KEY_TYPE
EOF
)

todo_fun_to_update=$(cat<<EOF
$todo_fun_to_update
fun_to_add=ssh_var_ini
fun_to_add=ssh_var_cal
ssh_secret_key_list=ssh_sec_key_list
ssh_secret_key_gen=ssh_sec_key_gen
ssh_key_check=ssh_sec_key_has
print_ssh_to_vm=ssh_to_vm_note_out
print_scp_to_vm=scp_to_vm_note_out
EOF
)

todo_lan_to_update=$(cat<<EOF
$todo_lan_to_update
lang_to_add=en
EOF
)

# out src index sh file path
#echo "${SRC_PATH}/index.sh"

# file usage
# sample/tom/app.add.sh
# sample/tom/app.add.sh "192.168.56.3" "root" "20-06-16" ~/.ssh/
# sample/tom/app.add.sh "$SSH_SERVER_IP" "$SSH_USER_NAME" "$SSH_KEY_FILE_NAME" "$SSH_PRI_KEY_FILE_PATH"